### ISSUE TITLE
Revise certification workflow

### ISSUE THREAD

**Issue by jadudm** [GSA-TTS/FAC#1495]
# At a glance

[comment]: # "Begin with a short summary so intent can be understood at a glance."

[comment]: # "In order to: some objective or value to be achieved"
[comment]: # "as a: stakeholder"
[comment]: # "I want: some new feature"

**In order to** finalize my audit
**as an** auditor/auditee
**I want** a certification workflow.

# Acceptance Criteria

We use [DRY](https://docs.behat.org/en/latest/user_guide/writing_scenarios.html#backgrounds) [behavior-driven development](https://en.wikipedia.org/wiki/Behavior-driven_development#Behavioral_specifications) wherever possible.

**Given**   a completed audit
**when**  I get to certification
...

[comment]: # "Each task should be a verifiable outcome"
```[tasklist]
### then...
- [x] Redesign the begin certification as a lock for certification page
- [x] Mock up the certification checklist in a "locked" state.
- [x] Develop content for the lock for certification page
- [x] Implement front-end "lock for certification" page
- [x] Update submission checklist to conditionally allow submission based on lock state
- [x] Update upload validation pathways (XLSX, PDF) to check for valid submission state
- [ ] https://github.com/GSA-TTS/FAC/issues/1516
- [x] Implement auditor certification workflow / page
- [x] Implement the auditee certification workflow / page
- [x] Implement final submission page
```

### Shepherd

[comment]: # "@ mention shepherds as we move across the board."

* Design shepherd: @lauraherring 
* Engineering shepherd: @heymatthenry 

# Background

[comment]: # "Any helpful contextual notes or links to artifacts/evidence, if needed"

We're now going to add a notion of forward-moving immutability to the certification process.

1. If this is a tribal submission, they must complete the public content attestation before certification locking. 
1. The "Ready to certify" page needs to be updated.
    1. We think we only want the Auditee Certifying Official to actually be able to push this button. It might be only they can see this link? (hm.)
    2. The user should be warned that locking for certification cannot be undone. They should make sure they have the final and correct versions of the workbooks and PDF in place.
    3. The button text should be clear that it is a locking action.
    4. Do we want a modal "Are you sure?" Possibly.
    5. Only the appropriate user(s) can hit the button successfully (in case someone shares a link to the page with a non-approved user)
    2. When the user "locks," we will set state to `READY_FOR_CERTIFICATION`. 
    3. Now, the `save()` method should reject all submissions with an error saying "the certification is locked, and no further form/PDF uploads are allowed." In this regard, users can still visit (via bookmark) a workbook upload page, but the submission should be impossible.
2. Users are redirected to the checklist page, and the checklist now grays out/whatever it does all of the submission links. Users cannot casually attempt to re-upload a component.
6. The link available is the auditor certifications step. Once that is complete, auditee certification unlocks.
7. Once both are done, final submission unlocks.

# Security Considerations

Required per [CM-4](https://nvd.nist.gov/800-53/Rev4/control/CM-4).

[comment]: # "Our SSP says 'The team ensures security implications are considered as part of the agile requirements refinement process by including a section in the issue template used as a basis for new work.'"
[comment]: # "Please do not remove this section without care."
[comment]: # "Note any security concerns that might be implicated in the change. 'None' is OK, but we must be explicit here."

No known security concerns; this falls within the same scope as other form submissions within the FAC.

# Sketch

https://www.figma.com/file/lnkXJTfKzD3YBhiJZ2QFEv/FAC-final?type=design&node-id=1-20991&mode=design&t=nP9jum5ysVEWRsga-4

---

<details>
  <summary>Process checklist</summary>
  
# Sketch

[comment]: # "Notes or a checklist reflecting our understanding of the selected approach"

- [ ] Design designs all the things
- [ ] Engineering engineers all the things

# Definition of Done

### If not likely to be important in the next quarter...
- [ ] Archived from the board

### Otherwise...

- [ ] Has a clear story statement
- [ ] Design or Engineering accepts that it belongs in their respective backlog

## Design Backlog

- [ ] Has clearly stated/testable acceptance criteria
- [ ] Meets the design Definition of Ready [citation needed]
- [ ] A design shepherd has been identified

## Design In Progress

- [ ] Meets the design Definition of Done [citation needed]

## Design Review Needed

- [ ] Necessary outside review/sign-off was provided

## Design Done

- [ ] Presented in a sprint review
- [ ] Includes screenshots or references to artifacts

### If no engineering is necessary
- [ ] Tagged with the sprint where it was finished
- [ ] Archived

## Engineering Backlog

- [ ] Has clearly stated/testable acceptance criteria
- [ ] Has a sketch or list of tasks
- [ ] Can reasonably be done in a few days (otherwise, split this up!)

## Engineering Available

- [ ] There's capacity in the `In Progress` column
- [ ] An engineering shepherd has been identified

## Engineering In Progress

- [ ] Meets acceptance criteria
- [ ] Meets [QASP conditions](https://derisking-guide.18f.gov/qasp/)

### If there's UI...
- [ ] Screen reader - Listen to the experience with a screen reader extension, ensure the information presented in order
- [ ] Keyboard navigation - Run through acceptance criteria with keyboard tabs, ensure it works. 
- [ ] Text scaling - Adjust viewport to 1280 pixels wide and zoom to 200%, ensure everything renders as expected. Document 400% zoom issues with USWDS if appropriate.

## Engineering Blocked

- [ ] Blocker removed/resolved

## Engineering Review Needed

- [ ] Outside review/sign-off was provided

## Engineering Done

- [ ] Presented in a sprint review
- [ ] Includes screenshots or references to artifacts
- [ ] Tagged with the sprint where it was finished
- [ ] Archived

</details>

**Comment by danswick** [comment]
>The link available is the auditor certifications step. Once that is complete, auditee certification unlocks.

@jadudm Does this mean the auditor must complete certification before the auditee can start certification? Or, more succinctly, does this mean there's an enforced order for certification? This came up during design sync and the prevailing opinion/hypothesis was that order probably doesn't matter, but it would be worth putting a fine point on it.

**Comment by jadudm** [comment]
I believe it is true that there is a state machine that enforces the order. 

To undo it would take work at this point; we believed that the auditor should go first, and the auditee last, because they have the ultimate responsibility for accounting for their expenditure of federal dollars.

So, yes. Once we lock for certification, the auditor link becomes available, but *not* the auditee link. (Why? Because, right now, the state machine will not allow the auditee to certify before the auditor is done.)

**Comment by JeanMarie-TTS** [comment]
@Leighdiddy, @Carley-Sullivan  , C-FAC has an auditor signature date and an auditee signature date on the form. Can these dates be different from the dates on which the certification is done in the FAC app? If so, do we need to add these fields to the form?

**Comment by jadudm** [comment]
I would say let's pull @timoballard into the conversation tomorrow.

I think it is reasonable to expect that the workflow/backend will record those dates when those forms are submitted. We don't want the user entering dates; we want to record the date/time when they do the cert.

So, we might need them to appear in our data, but I don't think we need to add them to a form?

**Comment by jadudm** [comment]
I think we've completed all of this work.
